src/serializer/ResidualReactElementVisitor.js

```diff
@@ -0,0 +1,126 @@
+    }
+
+    if (propsValue instanceof AbstractValue) {
+      // visit object, as it's going to be spread
+      this.residualHeapVisitor.visitValue(propsValue);
+    } else if (propsValue instanceof ObjectValue) {
+      if (propsValue.isPartialObject()) {
+        this.residualHeapVisitor.visitValue(propsValue);
+      } else {
+        // given that props is a concrete object, it should never be serialized
+        // as we'll be doing it directly with the ReactElementSerializer
+        // so we make the object as refusing serialization
+        propsValue.refuseSerialization = true;
+        for (let [propName, binding] of propsValue.properties) {
+          if (binding.descriptor !== undefined && propName !== "children") {
+            let propValue = getProperty(this.realm, propsValue, propName);
+            this.residualHeapVisitor.visitValue(propValue);
+          }
+        }
+      }
+    }
+  }
+
+  _visitReactElementChildren(reactElement: ObjectValue): void {
+    let propsValue = getProperty(this.realm, reactElement, "props");
+    if (!(propsValue instanceof ObjectValue)) {
+      return;
+    }
+    // handle children
+    if (propsValue.properties.has("children")) {
+      let childrenValue = getProperty(this.realm, propsValue, "children");
+      if (childrenValue !== this.realm.intrinsics.undefined && childrenValue !== this.realm.intrinsics.null) {
+        if (childrenValue instanceof ArrayValue && !childrenValue.intrinsicName) {
+          let childrenLength = getProperty(this.realm, childrenValue, "length");
+          let childrenLengthValue = 0;
+          if (childrenLength instanceof NumberValue) {
+            childrenLengthValue = childrenLength.value;
+            for (let i = 0; i < childrenLengthValue; i++) {
+              let child = getProperty(this.realm, childrenValue, "" + i);
+              invariant(
+                child instanceof Value,
+                `ReactElement "props.children[${i}]" failed to visit due to a non-value`
+              );
+              this.residualHeapVisitor.visitValue(child);
+            }
+          }
+        } else {
+          this.residualHeapVisitor.visitValue(childrenValue);
+        }
+      }
+    }
+  }
+
+  visitReactElement(reactElement: ObjectValue): void {
+    let typeValue = getProperty(this.realm, reactElement, "type");
+    let isReactFragment =
+      typeValue instanceof SymbolValue && typeValue === getReactSymbol("react.fragment", this.realm);
+
+    // we don't want to visit fragments as they are internal values
+    if (!isReactFragment) {
+      this.residualHeapVisitor.visitValue(typeValue);
+    }
+
+    this._visitReactElementAttributes(reactElement);
+    this._visitReactElementChildren(reactElement);
+
+    if (this.realm.react.output === "create-element" || isReactFragment) {
+      this.someReactElement = reactElement;
+    }
+    // check we can hoist a React Element
+    canHoistReactElement(this.realm, reactElement, this.residualHeapVisitor);
```

Comment:
It's confusing this looks like a boolean function call but apparently it has a mutation/side effect since return value is not used

Reply:
Will rename.

Reply:
Turns out we don't even need it! Tests all pass without it.